### PR TITLE
lsp: Provide data ref completions in rules

### PR DESCRIPTION
**Pull request by charlieegan3**
Fixes: https://github.com/StyraInc/regal/issues/753

https://github.com/StyraInc/regal/assets/1774239/fe4684ff-8b8a-4891-ad40-19d12df06937

### CODE REVIEW

File: internal/lsp/completions/providers/rulerefs.go [hunk 1]
```diff
@@ -33,108 +36,159 @@ func (*RuleFromImportedPackageRefs) Run(
...
+		for _, ref := range refs {
+			// we are not interested in packages here, only rules
 			if ref.Kind == types.Package {
 				continue
 			}
 
-			// don't suggest refs of "private" rules, even
-			// if only just by naming convention
+			// don't suggest refs of "private" rules, even if this
+			// is only just a naming convention
```
Comment (anderseknert) [inline internal/lsp/completions/providers/rulerefs.go:59]: Sneaky move, approving my PR only to correct me later hoping I wouldn't notice 😄

File: internal/lsp/completions/providers/rulerefs.go [hunk 2]
```diff
@@ -33,108 +36,159 @@ func (*RuleFromImportedPackageRefs) Run(
...
-	for file, refs := range c.GetAllFileRefs() {
-		if file == fileURI {
-			continue
-		}
-
-		for key, ref := range refs {
-			// we are not interested in packages here, only the rules
+	for uri, refs := range c.GetAllFileRefs() {
+		for _, ref := range refs {
+			// we are not interested in packages here, only rules
```
Comment (anderseknert) [inline internal/lsp/completions/providers/rulerefs.go:53]: I guess there are a few cases where you might want to refer to a package even in a rule body, like when doing dynamic policy composition:

```
some policy
some decision in data.policies[policy].deny
```
and so on... but I guess that's uncommon enough that it's still better to only suggest rules as a default.

File: internal/lsp/completions/providers/rulerefs.go [hunk 3]
```diff
@@ -33,108 +36,159 @@ func (*RuleFromImportedPackageRefs) Run(
...
+
+			switch {
+			case ref.Kind == types.ConstantRule:
+				symbol = completion.Constant
+				detail = "Constant Rule"
+			case ref.Kind == types.Function:
+				symbol = completion.Function
+				detail = "Function"
+			case ref.Description == "Partial":
+				detail = "Partial path suggestion, continue typing for more suggestions."
```
Comment (anderseknert) [inline internal/lsp/completions/providers/rulerefs.go:139]: cool!

**Review by anderseknert — APPROVED** [review]
Excellent! 👏

File: internal/lsp/completions/providers/rulerefs.go (hunk 1 above)
Comment (charlieegan3) [reply internal/lsp/completions/providers/rulerefs.go:59]: Ha, sorry...

File: internal/lsp/completions/providers/rulerefs.go (hunk 2 above)
Comment (charlieegan3) [reply internal/lsp/completions/providers/rulerefs.go:53]: Hmm, that is a thought. I guess this provide is called rulers and is already quite complex. I'd like to consider that for a new provider.

Issue here: https://github.com/StyraInc/regal/issues/771

**Pull request merged**